File: src/algorithms/strings/pattern-matching/boyer-moore-search/__tests__/boyer-moore-search.test.ts
import { describe, it, expect } from "vitest";
import { boyerMooreSearch } from "../sources/boyer-moore-search.ts?fn";

describe("boyerMooreSearch", () => {
  it("finds the pattern at the start of the text", () => {
    expect(boyerMooreSearch("ABCDEF", "ABC")).toBe(0);
  });

  it("finds the pattern in the middle of the text", () => {
    expect(boyerMooreSearch("ABAAABCD", "ABC")).toBe(4);
  });

  it("finds the pattern at the end of the text", () => {
    expect(boyerMooreSearch("XYZABC", "ABC")).toBe(3);
  });

  it("returns -1 when the pattern is not present", () => {
    expect(boyerMooreSearch("ABCDEFG", "XYZ")).toBe(-1);
  });

  it("handles a single-character pattern that exists", () => {
    expect(boyerMooreSearch("HELLO", "L")).toBe(2);
  });

  it("handles a single-character pattern that does not exist", () => {
    expect(boyerMooreSearch("HELLO", "Z")).toBe(-1);
  });

  it("returns 0 for an empty pattern", () => {
    expect(boyerMooreSearch("HELLO", "")).toBe(0);
  });

  it("handles text equal to the pattern", () => {
    expect(boyerMooreSearch("ABCD", "ABCD")).toBe(0);
  });

  it("returns -1 when pattern is longer than text", () => {
    expect(boyerMooreSearch("AB", "ABCD")).toBe(-1);
  });

  it("handles repeated characters with bad character skipping", () => {
    expect(boyerMooreSearch("AAAAABCD", "ABCD")).toBe(4);
  });

  it("finds a pattern that requires multiple shifts", () => {
    expect(boyerMooreSearch("GCATCGCAGAGAGTATACAGTACG", "GCAGAGAG")).toBe(5);
  });

  it("handles patterns with no repeated characters", () => {
    expect(boyerMooreSearch("ABCDEFGHIJK", "DEF")).toBe(3);
  });
});

File: src/algorithms/strings/pattern-matching/boyer-moore-search/__tests__/boyer-moore-search_test.go
package main

import "testing"

func TestBoyerMooreSearchPatternAtStart(t *testing.T) {
	if boyerMooreSearch("ABCDEF", "ABC") != 0 {
		t.Error("expected 0")
	}
}

func TestBoyerMooreSearchPatternInMiddle(t *testing.T) {
	if boyerMooreSearch("ABAAABCD", "ABC") != 4 {
		t.Error("expected 4")
	}
}

func TestBoyerMooreSearchPatternAtEnd(t *testing.T) {
	if boyerMooreSearch("XYZABC", "ABC") != 3 {
		t.Error("expected 3")
	}
}

func TestBoyerMooreSearchPatternNotFound(t *testing.T) {
	if boyerMooreSearch("ABCDEFG", "XYZ") != -1 {
		t.Error("expected -1")
	}
}

func TestBoyerMooreSearchSingleCharFound(t *testing.T) {
	if boyerMooreSearch("HELLO", "L") != 2 {
		t.Error("expected 2")
	}
}

func TestBoyerMooreSearchSingleCharNotFound(t *testing.T) {
	if boyerMooreSearch("HELLO", "Z") != -1 {
		t.Error("expected -1")
	}
}

func TestBoyerMooreSearchEmptyPattern(t *testing.T) {
	if boyerMooreSearch("HELLO", "") != 0 {
		t.Error("expected 0 for empty pattern")
	}
}

func TestBoyerMooreSearchTextEqualsPattern(t *testing.T) {
	if boyerMooreSearch("ABCD", "ABCD") != 0 {
		t.Error("expected 0")
	}
}

func TestBoyerMooreSearchPatternLongerThanText(t *testing.T) {
	if boyerMooreSearch("AB", "ABCD") != -1 {
		t.Error("expected -1")
	}
}

func TestBoyerMooreSearchRepeatedChars(t *testing.T) {
	if boyerMooreSearch("AAAAABCD", "ABCD") != 4 {
		t.Error("expected 4")
	}
}

func TestBoyerMooreSearchMultipleShifts(t *testing.T) {
	if boyerMooreSearch("GCATCGCAGAGAGTATACAGTACG", "GCAGAGAG") != 5 {
		t.Error("expected 5")
	}
}

func TestBoyerMooreSearchNoRepeatedChars(t *testing.T) {
	if boyerMooreSearch("ABCDEFGHIJK", "DEF") != 3 {
		t.Error("expected 3")
	}
}
